### ISSUE TITLE
AbstractVar property: interaction bug with static type-checker

### ISSUE THREAD

**Issue by EtaoinWu** [patrick-kidger/equinox#882]
The following code:

```python
import equinox as eqx

class A(eqx.Module):
    x: eqx.AbstractVar[int]

class B(A):
    y: int

    @property
    def x(self):
        return self.y
    
b = B(y=3)
b.x
```

runs without a problem, but gives error in pyright and mypy. Pyright complains

```plain
⮞ mypy ./test.py
test.py:9: error: Dataclass attribute may only be overridden by another attribute  [misc]
test.py:13: error: Missing positional argument "x" in call to "B"  [call-arg]
Found 2 errors in 1 file (checked 1 source file)
```

and mypy complains

```plain
⮞ pyright ./qwq.py
/path/to/test.py
  /path/to/test.py:10:9 - error: "x" overrides symbol of same name in class "A"
    "property" is not assignable to "int" (reportIncompatibleVariableOverride)
  /path/to/test.py:13:5 - error: Argument missing for parameter "x" (reportCallIssue)
2 errors, 0 warnings, 0 informations
```

My pylance in vscode somehow only reports the second error ("x" missing) but not the first (property cannot override int).

I believe that `AbstractVar property` is an intentional feature, as the doc says

> An `AbstractVar[T]` must be overridden by an attribute annotated with `AbstractVar[T]`, `AbstractClassVar[T]`, `ClassVar[T]`, `T`, or **a property returning `T`**.

I suggest this to be fixed by `_ModuleMeta` adding something (to its `__new__`?) at typechecking time to modify the `__annotations__` of its initializer.

**Comment by patrick-kidger** [comment]
Yup, implementing an `AbstractVar` with a `property` is an intentional feature.

That said, I don't know of a way to make this static-type-checking compatible. If you think you have a way to do that then I'd be happy to see a PR doing so!

**Comment by EtaoinWu** [comment]
This code:

```python
import equinox as eqx

class A(eqx.Module):
    x: eqx.AbstractVar[int] = eqx.field(init=False)

class B(A):
    y: int

    @property
    def x(self):
        return self.y
    
class C(A):
    x: int
    
b = B(y=3)
print(b.x)
c = C(x=3)
print(c.x)
```

makes static typecheckers happy, but fails in `_better_abstract.py` with `Abstract attribute x cannot have value`. Could we allow `field(init=False)` in the value just to make pyright and mypy work?

**Comment by patrick-kidger** [comment]
So it's quite important that this not actually create a dataclass field -- which the above snippet does do -- as we may later want to implement an `AbstractVar` as a `@property`.

You could surrounding these in an `if TYPE_CHECKING` block, perhaps. I'm not sure if there's some ergonomic alternative we could find for Equinox itself.
